### DISCUSSION TITLE
Build v2rayNG App with github action

### DISCUSSION THREAD

**Discussion by ghost** [2dust/v2rayNG#2527] [Show and tell]
Hey everyone,

I wanted to share that you can build the v2rayNG app with ease by using GitHub Action! 
Here are the steps to get started:
1. [fork my repo](https://github.com/oXIIIo/v2rayNG/fork)
2. Go to the "Settings" tab of your repository, and then click on "Secrets."
3. Add add the following **GitHub Secrets** to your repository:
   - `SIGNING_KEY`: Your signing key (Keystore) in base64-encoded format.
   - `SIGNING_KEY_ALIAS`: The alias for your signing key.
   - `SIGNING_KEY_PASSWORD`: The password for your signing key.
   - `SIGNING_STORE_PASSWORD`: The store password for your signing key.
4. To trigger the GitHub Actions workflow manually, navigate to the "Actions" tab, select the workflow you want to run, and click the "Run workflow" button. This will initiate the build process.
Additionally, please make sure to provide the input data for XRAY_CORE_VERSION and V2RAYNG_VERSION when running the workflow.
   - `XRAY_CORE_VERSION`: The version of XRay Core to use for the build e.g `v1.8.4`.
   - `V2RAYNG_VERSION`: The version of v2rayNG (you can pass anything you want , it's not important).
5. Once the workflow completes, the built app will be uploaded to the GitHub Releases section of your repository. You can find it there and download the APK for installation on your Android device.

That's it! You should now have a working v2rayNG app built from the latest source code. Feel free to explore and customize it as needed.

Repository Link: https://github.com/oXIIIo/v2rayNG

**Comment by a-pav** [comment]
> 3. Add add the following GitHub Secrets to your repository: 

How can I generate these 4 `SIGNING_` fields?
I don't have Android Studio right now. I'm looking for the easiest way to build v2rayNG.

**Reply by ghost** [reply]
**1. Download the Java Development Kit (JDK):**

Head over to the Oracle website and download the latest version of the Java Development Kit (JDK). You can find it on the Java download page: [https://www.oracle.com/java/technologies/downloads/](https://www.oracle.com/java/technologies/downloads/).

**2. Generate a Keystore:**

Once you've downloaded and installed the JDK, unzip the file. Navigate to the `bin` folder within the JDK installation directory. Open a terminal window in this folder (right-click and choose "Open in Terminal" on macOS/Linux or use the command prompt on Windows).

Now, use the `keytool` command to generate a keystore. Here's the command structure with explanations for each option:

```
keytool -genkey -v [-keystore <keystore_name>] [-alias <alias_name>] [-storepass <store_password>] [-keypass <key_password>] [-validity <validity_period_in_days>] [-keyalg <key_algorithm>] [-keysize <key_size>]
```

**Options:**

* `-genkey`: This option tells `keytool` to generate a new key pair.
* `-v`: This enables verbose output, providing more details about the generated key.
* `-keystore <keystore_name> (optional)`: This specifies the name of the keystore file to be created. The default is `keystore.jks`.
* `-alias <alias_name> (optional)`: This defines the alias (name) for the key entry within the keystore.
* `-storepass <store_password> (optional)`: This sets the password required to access the entire keystore.
* `-keypass <key_password> (optional)`: This sets the password specifically for the generated key pair. It's recommended to use a different password than the store password.
* `-validity <validity_period_in_days> (optional)`: This defines the validity period of the generated key pair in days. The default is 365 days (one year).
* `-keyalg <key_algorithm> (optional)`: This specifies the algorithm used to generate the key pair. The default is RSA, which is recommended for most cases.
* `-keysize <key_size> (optional)`: This defines the key size in bits. Common choices are 2048 or 4096. A larger key size provides stronger security but can also impact performance.

**Example Usage:**

```
keytool -genkey -v -alias mykey -storepass mystorepass -keypass mykeypass -validity 3650 -keyalg RSA -keysize 2048
```

This command will generate a keystore named `mykey.jks` containing a key entry with the alias `mykey`. It will use passwords `mystorepass` for the keystore and `mykeypass` for the key itself. The key will be valid for 3650 days (10 years) and use the RSA algorithm with a key size of 2048 bits.

**Remember to replace `<keystore_name>`, `<alias_name>`, `<store_password>`, `<key_password>`, `<validity_period_in_days>`, `<key_algorithm>`, and `<key_size>` with your desired values.**

**3. Upload the Keystore and Generate Base64 Encoded Key:**

Upload the generated `keystore.jks` file to your VPS using a secure method like SFTP. Once uploaded, navigate to the directory containing the keystore in your SSH terminal. Run the following command to convert the keystore to base64 encoded format:

```
openssl base64 < keystore.jks | tr -d '\n' | tee keystore_base64.txt
```

This command:

* Uses `openssl base64` to convert the binary keystore file to base64 encoded format.
* Pipes the output to `tr -d '\n'` to remove any newline characters.
* Finally, saves the encoded key data to a file named `keystore_base64.txt`.

**Reply by a-pav** [reply]
Thank you.
Before I had your response, I did it with `keytool -list -rfc -keystore ./myks.keystore` then copy-pasted the key as `SIGNING_KEY` in github actions secrects and it worked OK. [link.](https://stackoverflow.com/a/23302895/5356496) (I've included the `-----BEGIN/END CERTIFICATE-----` lines too). Next time I might use `openssl` method as you suggested, but right now I see that it produces a different output. I'm not sure how github can work with both.
